assimilation_code/modules/io/state_structure_mod.f90

```diff
@@ -765,114 +769,157 @@ subroutine load_common_cf_conventions(domain)
 do ivar = 1, nvars
    var_name = domain%variable(ivar)%varname
 
-   call nc_check(nf90_inq_varid(ncid, trim(var_name), VarID), &
-            'load_common_cf_conventions', 'inq_varid '//trim(var_name))
+   ret = nf90_inq_varid(ncid, trim(var_name), VarID)
+   call nc_check(ret, routine, 'inq_varid '//trim(var_name))
 
    ! If the short_name, long_name and/or units attributes are set, get them.
    ! They are not REQUIRED by DART but are nice to keep around if they are present.
 
    if( nf90_inquire_attribute(    ncid, VarID, 'long_name') == NF90_NOERR ) then
-      call nc_check( nf90_get_att(ncid, VarID, 'long_name' , cf_long_name), &
-                     'load_common_cf_conventions', 'get_att long_name '//trim(var_name))
+      ret = nf90_get_att(ncid, VarID, 'long_name' , cf_long_name)
+      call nc_check(ret, routine, 'get_att long_name '//trim(var_name))
       domain%variable(ivar)%io_info%long_name = cf_long_name
    endif
 
    if( nf90_inquire_attribute(    ncid, VarID, 'short_name') == NF90_NOERR ) then
-      call nc_check( nf90_get_att(ncid, VarID, 'short_name' , cf_short_name), &
-                     'load_common_cf_conventions', 'get_att short_name '//trim(var_name))
+      ret = nf90_get_att(ncid, VarID, 'short_name' , cf_short_name)
+      call nc_check(ret, routine, 'get_att short_name '//trim(var_name))
       domain%variable(ivar)%io_info%short_name = cf_short_name
    endif
 
    if( nf90_inquire_attribute(    ncid, VarID, 'units') == NF90_NOERR )  then
-      call nc_check( nf90_get_att(ncid, VarID, 'units' , cf_units), &
-                  'load_common_cf_conventions', 'get_att units '//trim(var_name))
+      ret = nf90_get_att(ncid, VarID, 'units' , cf_units)
+      call nc_check(ret, routine, 'get_att units '//trim(var_name))
       domain%variable(ivar)%io_info%units = cf_units
    endif
 
-   ! Saving any FillValue, missing_value attributes.
+   ! Saving any FillValue, missing_value attributes ...
+   ! Also stuff them into the R8 slots to facilitate simpler clamp_variable
+   ! implementation.
 
    var_xtype = domain%variable(ivar)%io_info%xtype
    select case (var_xtype)
       case ( NF90_INT )
-          if (nf90_get_att(ncid, NF90_GLOBAL, '_FillValue', cf_spvalINT) == NF90_NOERR) then
+          ret = nf90_get_att(ncid, NF90_GLOBAL, '_FillValue',    cf_spvalINT)
+          if (ret == NF90_NOERR) then
              domain%variable(ivar)%io_info%spvalINT       = cf_spvalINT
-             domain%variable(ivar)%io_info%has_missing_value = .true.
+             domain%variable(ivar)%io_info%spvalR8             = real(cf_spvalINT,r8)
```

Comment:
why is there a spvalR8 and a missing_value in the io_info struct?  when will they be different?  and again - i don't see why the code needs to have per-type missing values since the only use will be when they're converted to R8.